statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,2146 @@
+                                order+1)).fit().resid
+
+    def resid(y, x):
+        if x.size == 0:
+            return y
+        r = y - np.dot(x, np.dot(np.linalg.pinv(x), y))
+        return r
+
+    nobs, neqs = endog_tot.shape
+
+    # why this?  f is detrend transformed series, det_order is detrend data
+    if det_order > -1:
+        f = 0
+    else:
+        f = det_order
+
+    endog_tot = detrend(endog_tot, det_order)
+    dx = tdiff(endog_tot, 1, axis=0)
+    z = mlag(dx, k_ar)
+    z = trimr(z, k_ar, 0)
+    z = detrend(z, f)
+
+    dx = trimr(dx, k_ar, 0)
+
+    dx = detrend(dx, f)
+    r0t = resid(dx, z)
+    lx = lag(endog_tot, k_ar)
+    lx = trimr(lx, 1, 0)
+    dx = detrend(lx, f)
+    rkt = resid(dx, z)  # level on lagged diffs
+    skk = np.dot(rkt.T, rkt) / rkt.shape[0]
+    sk0 = np.dot(rkt.T, r0t) / rkt.shape[0]
+    s00 = np.dot(r0t.T, r0t) / r0t.shape[0]
+    sig = np.dot(sk0, np.dot(inv(s00), sk0.T))
+    tmp = inv(skk)
+    au, du = np.linalg.eig(np.dot(tmp, sig))  # au is eval, du is evec
+
+    temp = inv(np.linalg.cholesky(np.dot(du.T, np.dot(skk, du))))
+    dt = np.dot(du, temp)
+
+    # JP: the next part can be done much  easier
+    auind = np.argsort(au)
+    aind = np.flipud(auind)
+    a = au[aind]
+    d = dt[:, aind]
+
+    #  Compute the trace and max eigenvalue statistics
+    lr1 = np.zeros(neqs)
+    lr2 = np.zeros(neqs)
+    cvm = np.zeros((neqs, 3))
+    cvt = np.zeros((neqs, 3))
+    iota = np.ones(neqs)
+    t, junk = rkt.shape
+    for i in range(0, neqs):
+        tmp = trimr(np.log(iota-a), i, 0)
+        lr1[i] = -t * np.sum(tmp, 0)
+        lr2[i] = -t * np.log(1-a[i])
+        cvm[i, :] = c_sja(neqs - i, det_order)
+        cvt[i, :] = c_sjt(neqs - i, det_order)
+        aind[i] = i
+
+    result = Holder()
+    # estimation results, residuals
+    result.rkt = rkt
+    result.r0t = r0t
+    result.eig = a
+    result.evec = d
+    result.lr1 = lr1
+    result.lr2 = lr2
+    result.cvt = cvt
+    result.cvm = cvm
+    result.ind = aind
+    result.meth = 'johansen'
+
+    return result
+
+
+class VECM(tsbase.TimeSeriesModel):
+    """
+    Fit a VECM process
+    .. math:: \\Delta y_t = \\Pi y_{t-1} + \\Gamma_1 \\Delta y_{t-1} + \\ldots + \\Gamma_{k_ar-1} \\Delta y_{t-k_ar+1} + u_t
+    where
+    .. math:: \\Pi = \\alpha \\beta'
+    as described in chapter 7 of [1]_.
+
+    Parameters
+    ----------
+    endog_tot : array-like (nobs_tot x neqs)
+        2-d endogenous response variable.
+    exog: ndarray (nobs_tot x neqs) or None
+        Deterministic terms outside the cointegration relation.
+    exog_coint: ndarray (nobs_tot x neqs) or None
+        Deterministic terms inside the cointegration relation.
+    dates : array-like of datetime, optional
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
+        information.
+    freq : str, optional
+        See :class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel` for more
```

Comment:
I forgot `freq` and `missing` which were part of the scaffold I started from. After the last rebase I get warnings related to `freq`. Shall I just use `freq` instead of `seasons`?